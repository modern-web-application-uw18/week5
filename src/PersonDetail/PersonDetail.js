import React, { Component } from 'react';
import './PersonDetail.css';
import PropTypes from 'prop-types';

class PersonDetail extends Component {
  constructor(props) {
    super(props);
    this.state = {
      person: {}
    }
  }

  componentDidMount() {
    const peopleId = this.props.match.params.peopleId;
    fetch(`https://swapi.dev/api/people/${peopleId}/`)
      .then(response => response.json())
      .then(data => {
        this.setState((prevState, props) => {
          return {
            person: data
          };
        });
      })
      .catch(error => console.log(error));
  };

  getFilmList(films) {
    const filmNames = ["A New Hope", "The Empire Strikes Back", "Return of the Jedi", "The Phantom Menace", "Attack of the Clones", "Revenge of the Sith", "The Force Awakens", "The Last Jedi"];
    let finalFilmList = '';
    for (let i = 0; i < films.length; i++){
      const url = films[i];
      let filmNum = url.split('/').reverse()[1];
      filmNum = filmNum - 1;
      finalFilmList += filmNames[filmNum];
      if (i < (films.length - 1)) {
        finalFilmList += ", ";
      }
    }
    return (
      <span>
        {finalFilmList}
      </span>
    );
  };
  // // Can't do below as return statement in getFilmList
  // // tells me that finalFilmList.map is not a function;
  // // tried finalFilmList.forEach and got the same error.
  //
  // {finalFilmList.map((finalFilmList, idx) => {
  //    return <span>{finalFilmList}, </span>
  // })}

  getHomePlanet(planet) {
    const planetNames = ["Tatooine","Alderaan","Yavin IV","Hoth","Dagobah","Bespin","Endor","Naboo","Coruscant", "Kamino", "Geonosis", "Utapau", "Mustafar", "Kashyyyk", "Polis Massa", "Mygeeto", "Felucia", "Cato Neimoidia", "Saleucami", "Stewjon", "Eriadu", "Corellia", "Rodia", "Nal Hutta", "Dantooine", "Bestine IV", "Ord Mantell", "unknown", "Trandosha", "Socorro", "Mon Cala", "Chandrila", "Sullust", "Toydaria", "Malastare", "Dathomir", "Ryloth", "Aleen Minor", "Vulpter", "Troiken", "Tund", "Haruun Kal", "Cerea", "Glee Anselm", "Iridonia", "Tholoth", "Iktotch", "Quermia", "Dorin", "Champala", "Mirial", "Serenno", "Concord Dawn","Zolan","Ojom","Skako","Muunilinst","Shili","Kalee","Umbara","Jakku"]
    const url = planet;
    let planetNum = url.split('/').reverse()[1];
    planetNum = planetNum - 1;
    const finalPlanet = planetNames[planetNum];
    return (
      <span>
        {finalPlanet}
      </span>
    );
  };

  render() {
    const { name, height, mass, hair_color, skin_color, eye_color, birth_year, gender, homeworld } = this.state.person;
    const films = this.state.person.films;
    const peopleIdNum = this.props.match.params.peopleId;
    const backgroundStyles = {
        backgroundImage: 'url(\'/characters/character-' + peopleIdNum + '.jpg\')',
    };
    return (
      <div className="page-content">
        <section className="person-detail">
          <div className="content">
            <h1>{name}</h1>
            <aside className="stats">
              <h2>Vital Stats</h2>
              <div className="stats--shape">
                <div className="stat--chart">
                  <dl><dt>Height:</dt><dd>{height} cm</dd></dl>
                  <dl><dt>Mass:</dt><dd>{mass} kg</dd></dl>
                  <dl><dt>Hair color:</dt><dd>{hair_color}</dd></dl>
                  <dl><dt>Skin color:</dt><dd>{skin_color}</dd></dl>
                </div>
                <div className="stat--chart">
                  <dl><dt>Eye color:</dt><dd>{eye_color}</dd></dl>
                  <dl><dt>Birth year:</dt><dd>{birth_year}</dd></dl>
                  <dl><dt>Gender:</dt><dd>{gender}</dd></dl>
                  <dl><dt>Homeworld:</dt>{homeworld ? this.getHomePlanet(homeworld) : "unknown"}<dd></dd></dl>
                </div>
                <div className="stat--notes">
                  <p> Appears in: <span className="emph">{films ? this.getFilmList(films) : "none" }</span></p>
                </div>
              </div>
            </aside>
          </div>

          <div className="media" style={backgroundStyles}>
            <div className="styleized-font">{name}</div>
            <div className="styleized-font">{name}</div>
          </div>
        </section>
      </div>
    )
  }
}

PersonDetail.propTypes = {
  person: PropTypes.shape({
    name: PropTypes.string.isRequired,
    height: PropTypes.string,
    mass: PropTypes.string,
    hair_color: PropTypes.string,
    skin_color: PropTypes.string,
    eye_color: PropTypes.string,
    birth_year: PropTypes.string,
    gender: PropTypes.string,
    homeworld: PropTypes.string,
    films: PropTypes.arrayOf(PropTypes.string)
  })
};

export default PersonDetail;

