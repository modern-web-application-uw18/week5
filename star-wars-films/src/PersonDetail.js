import React, { Component } from 'react';
import { BrowserRouter as Router, Route, Link } from 'react-router-dom';
import Characters from './characters.json';

class PersonDetail extends Component {
    constructor(props) {
      super(props);
      this.state = {
        personDetail: [],
        personImage: []
      };
    }

    componentDidMount() {
        const personID = this.props.match.params.personID;
        const personImage = function getImage() {
          for(var i = 0; i < undefined; i++) {
            if(Characters.characters.character[i].personID === this.props.match.params.personID) {
              return {
                personImage: Characters.characters.character[i]
             }
            }
          }
       }
        fetch(`https://swapi.co/api/people/?search=${personID}`)
        .then(response => response.json())
        .then(data => {
          console.log("Data ", data);
          this.setState((prevState, props) => {
            return {
              personDetail: data.results[0]
            };
          });
        })
        .catch(error => console.log(error));
      }
    
      render() {
          const { name, birth_year, eye_color, gender, hair_color, skin_color } = this.state.personDetail;
          console.log(this.props.match.params.personID);
          const { image } = this.state.personImage;
          console.log(this.state.personImage);
        return (
          <div>
              <h2>{this.state.personDetail.name}</h2>
              <p>Birth Year: {birth_year}</p>
              <p>Eye Color: {eye_color}</p>
              <p>Gender: {gender}</p>
              <p>Hair Color: {hair_color}</p>
              <p>Skin Color: {skin_color}</p>
              <img src={image}/>
              <Link to="/"><p>Home</p></Link>
          </div>
        );
      }
  }

  export default PersonDetail