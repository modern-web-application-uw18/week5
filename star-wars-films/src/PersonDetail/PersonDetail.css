.person-detail {
  display: flex;
  flex-direction: row;
  align-items: stretch;
  margin-bottom: 4rem
}

.person-detail .media {
  background-color: #b0b0b0;
  position: relative;
  background-position: center right;
  background-size: cover;
  flex-basis: 60%;
  overflow: hidden;
  flex-shrink: 1;
}

.person-detail .media div:first-child {
  position: absolute;
  top: .5rem;
  left: 5rem;
}
.person-detail .media div:nth-child(2) {
  position: absolute;
  bottom: .5rem;
  right: 8rem;
}

.person-detail .media::after {
  content: '';
  width: 0;
  height: 0;
  position: absolute;
  bottom: 0;
  border-left: 8rem solid transparent;
  border-right: 8rem solid transparent;
  border-bottom: 8rem solid gainsboro;
  right: -8rem;
}

.person-detail .media::before {
  content: '';
  width: 0;
  height: 0;
  position: absolute;
  top: 0;
  border-left: 4rem solid transparent;
  border-right: 4rem solid transparent;
  border-top: 4rem solid gainsboro;
  left: -4rem;
}

.person-detail .stats {
  background-color: white;
  position: relative;
  margin-right: -2rem;
  z-index: 6;
}

.person-detail .stats .stats--shape::after {
  content: '';
  width: 0;
  height: 0;
  position: absolute;
  bottom: 0;
  border-left: 4rem solid transparent;
  border-right: 4rem solid transparent;
  border-bottom: 4rem solid gainsboro;
  left: -4rem;
}

.person-detail .stats::before {
  content: '';
  width: 0;
  height: 0;
  position: absolute;
}

.person-detail .stats .stats--shape {
  background-color: white;
  height: auto;
  padding: 0 2rem 2rem 3rem;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  flex-wrap: wrap;
}

.person-detail .content h1 {
  font-family: 'Glegoo', serif;
  font-size: 2rem;
  color: #585858;
}

.person-detail .stats h2 {
  font-family: 'Glegoo', serif;
  font-size: 1.2rem;
  font-weight: 600;
}

.person-detail .stats--shape dt {
  font-weight: 600;
  margin-right: .2rem;
  color: #585858;
}

.person-detail .stats--shape dl {
  padding: .2rem;
  margin-bottom: .4rem;
  background-color: #f8f8f8;
}

.person-detail .stats--shape > .stat--chart {
  width: 49%;
  padding-top: .4rem;
}

.person-detail .stats--shape > .stat--notes {
  width: 100%;
  flex-basis: 100%;
  margin-top: .8rem;
  padding-bottom: .5rem;
  border-bottom: 1px solid gainsboro;
}

.person-detail .stats--shape > .stat--notes p {
  font-size: 1rem;
  line-height: 1.2;
}

.person-detail .stats--shape dt,
.person-detail .stats--shape dd {
  display: inline-block;
  position: relative;
  width: auto;
}

@media screen and (max-width: 767px) {
  .person-detail {
    flex-direction: column-reverse;
    margin: 0 1rem 1rem 1rem;
  }
  .person-detail .content {
    flex-basis: auto;
  }
  .person-detail .content h1 {
    margin: 1rem 0 1rem 1rem;
  }
  .person-detail .stats {
    margin-left: 0;
    margin-right: 0;
    width: 100%;
    padding: 2rem 0 0 0;
    overflow: auto;
  }
  .person-detail .stats::before {
    top: 0;
    border-left: 4rem solid transparent;
    border-right: 4rem solid transparent;
    border-top: 4rem solid gainsboro;
    right: -4rem;
  }
  .person-detail .media {
    height: 16rem;
  }
  .person-detail .stats .stats--shape {
    width: 100%;
    position: relative;
  }
  .person-detail .stats h2 {
    margin-top: 0;
    margin-left: 3rem;
    margin-bottom: 1rem;
  }
}

@media screen and (min-width: 768px) {
  .person-detail {
    flex-direction: row;
  }
  .person-detail .content {
    flex-basis: 40%;
  }
  .person-detail .content h1 {
    margin: 1rem 0 1rem 4rem;
  }
  .person-detail .stats {
    margin-left: 2rem;
    margin-right: -2rem;
    padding: 3rem 2rem 4rem 2rem;
    overflow: visible;
  }
  .person-detail .stats::before {
    top: 2rem;
    border-left: 4rem solid transparent;
    border-right: 4rem solid transparent;
    border-bottom: 4rem solid white;
    right: -6rem;
    transform: rotate(90deg);
  }
  .person-detail .media {
    height: 24rem;
  }
  .person-detail .stats .stats--shape {
    width: calc(100% + 4rem);
    right: -4rem;
    position: absolute;
    top: 4rem;
  }
  .person-detail .stats h2 {
    margin-top: -1rem;
    margin-left: 1rem;
  }
}
