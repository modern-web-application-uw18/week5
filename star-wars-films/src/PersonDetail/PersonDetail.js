import React, { Component } from 'react';

class PersonDetail extends Component {
  constructor(props) {
    super(props);
    this.state = {
      people: {}
    }
  }

  componentDidMount() {
    const peopleId = this.props.match.params.peopleId;
    console.log(peopleId);
    fetch(`https://swapi.co/api/people/${peopleId}/`)
      .then(response => response.json())
      .then(data => {
        console.log(data);
        this.setState((prevState, props) => {
          return {
            people: data
          };
        });
      })
      .catch(error => console.log(error));
  };

  render() {
    const { name, height, mass, hair_color, skin_color, eye_color } = this.state.people
    return (
      <div>
        <h1>{name}</h1>
        <dl><dt>height</dt><dd>{height}</dd></dl>
        <dl><dt>mass</dt><dd>{mass}</dd></dl>
        <dl><dt>hair color</dt><dd>{hair_color}</dd></dl>
        <dl><dt>skin color</dt><dd>{skin_color}</dd></dl>
        <dl><dt>eye color</dt><dd>{eye_color}</dd></dl>
      </div>
    )
  }
}

export default PersonDetail;
