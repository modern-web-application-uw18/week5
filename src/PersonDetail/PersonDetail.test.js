import React from 'react';
import PersonDetail from './PersonDetail';
import ReactTestRenderer from 'react-test-renderer';

describe('PersonDetail component', () => {
  it('should render', () => {
    const match = { params: { peopleId: 1 } }
    const component = ReactTestRenderer.create(
      <PersonDetail match={match} />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
