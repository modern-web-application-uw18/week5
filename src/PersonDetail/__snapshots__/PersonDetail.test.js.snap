// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonDetail component should render 1`] = `
<div
  className="page-content"
>
  <section
    className="person-detail"
  >
    <div
      className="content"
    >
      <h1 />
      <aside
        className="stats"
      >
        <h2>
          Vital Stats
        </h2>
        <div
          className="stats--shape"
        >
          <div
            className="stat--chart"
          >
            <dl>
              <dt>
                Height:
              </dt>
              <dd>
                 cm
              </dd>
            </dl>
            <dl>
              <dt>
                Mass:
              </dt>
              <dd>
                 kg
              </dd>
            </dl>
            <dl>
              <dt>
                Hair color:
              </dt>
              <dd />
            </dl>
            <dl>
              <dt>
                Skin color:
              </dt>
              <dd />
            </dl>
          </div>
          <div
            className="stat--chart"
          >
            <dl>
              <dt>
                Eye color:
              </dt>
              <dd />
            </dl>
            <dl>
              <dt>
                Birth year:
              </dt>
              <dd />
            </dl>
            <dl>
              <dt>
                Gender:
              </dt>
              <dd />
            </dl>
            <dl>
              <dt>
                Homeworld:
              </dt>
              unknown
              <dd />
            </dl>
          </div>
          <div
            className="stat--notes"
          >
            <p>
               Appears in: 
              <span
                className="emph"
              >
                none
              </span>
            </p>
          </div>
        </div>
      </aside>
    </div>
    <div
      className="media"
      style={
        Object {
          "backgroundImage": "url('/characters/character-undefined.jpg')",
        }
      }
    >
      <div
        className="styleized-font"
      />
      <div
        className="styleized-font"
      />
    </div>
  </section>
</div>
`;
